### PR TITLE
releng - pytest-terraform 0.5.1 + developer documentation

### LINKED ISSUE #6130: docs - developer docs on how to use pytest-terraform fixtures
workflow around replay=False, record_flight_data and switching to replay_flight_data .. biggest bits are not using unit test class derived test class and using `test` fixture to get c7n test api.

### PR DESCRIPTION
**Pull request by marcoceppi**
This includes a bugfix to pytest-terraform and updates the developer testing documentation to include instructions on writing test using pytest-terraform

fixes #6130

### CODE REVIEW

File: docs/source/developer/tests.rst [hunk 1]
```diff
@@ -117,27 +126,170 @@ test is run while `record_flight_data` is in place.
...
 
     import time
 
     ...
 
-    if self.recording:
+    if test.recording:
       time.sleep(10)
+
+Writing Terraform Bootstrapped Tests for AWS Resources
```
Comment (kapilt) [inline docs/source/developer/tests.rst:153]: please remove the aws specific labeling here

File: docs/source/developer/tests.rst [hunk 2]
```diff
@@ -117,27 +126,170 @@ test is run while `record_flight_data` is in place.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Note: These instructions are only for recording data in AWS. For Azure, GCP, or
```
Comment (kapilt) [inline docs/source/developer/tests.rst:156]: can we drop this comment, we'll add examples of the other providers.

File: docs/source/developer/tests.rst [hunk 3]
```diff
@@ -42,8 +42,18 @@ directly from Microsoft for any hypervisor.
 Decorating tests
 ~~~~~~~~~~~~~~~~
 
-The ``functional`` decorator marks tests that don't require any pre-existing
-AWS context, and can therefore be run cleanly against live AWS.
+The ``terraform`` decorator marks tests that don't require any pre-existing context and can be run cleanly against live AWS accounts.
```
Comment (kapilt) [inline docs/source/developer/tests.rst:45]: this is a little too backwards looking without explaining how to use the terraform decorator, the example below is a better worked example. i would just remove this section or move the section on using the terraform decorator up.

File: docs/source/developer/tests.rst [hunk 4]
```diff
@@ -117,27 +126,170 @@ test is run while `record_flight_data` is in place.
...
+
+Writing Terraform Bootstrapped Tests for AWS Resources
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Note: These instructions are only for recording data in AWS. For Azure, GCP, or
+other cloud providers, see the corresponding documentation for information on how
+to record data there.
+
+The `pytest-terraform <https://github.com/cloud-custodian/pytest-terraform>`_ plugin
+is used to streamline the creation and removal of AWS resources using Terraform during
```
Comment (kapilt) [inline docs/source/developer/tests.rst:161]: s/aws/cloud

File: docs/source/developer/tests.rst [hunk 5]
```diff
@@ -117,27 +126,170 @@ test is run while `record_flight_data` is in place.
...
+When invoked, this terraform module will create a file `bar.txt` with the
+contents ``bar!``.
+
+When first creating a test, explicitly set the `replay` parameter to ``False``.
+This will force terraform to run on each invocation of the test and perform
+the flight recording function.
+
+Given the above example terraform file, the following code example is used to
+access data from the terraform decorator, passed in as the same name as the
+first parameter, in this case `file_example`
```
Comment (kapilt) [inline docs/source/developer/tests.rst:186]: this is missing the caveat wrt to unit test class derivatives, and using the `test` fixture for access to our existing c7n test api.

File: docs/source/developer/tests.rst [hunk 6]
```diff
@@ -117,27 +126,170 @@ test is run while `record_flight_data` is in place.
...
+
+Now when the test is run it will use the data previously recorded terraform
+resources and not run terraform directly. When committing your test, don't
+forget to include the `tests/terraform/file_example` directory!
+
+Controlling Resource Cleanup
+============================
+
+By default, pytest-terraform will run ``terraform destroy`` after each wrapped
+test function. For majority of use cases this will likely be okay. If terraform
```
Comment (kapilt) [inline docs/source/developer/tests.rst:234]: actually it scoped to session

Comment (kapilt) [inline docs/source/developer/tests.rst:234]: explain why it wouldn't be.. ie. the policy test execution would delete provisioned resources from terraform.

File: docs/source/developer/tests.rst (hunk 2 above)
Comment (marcoceppi) [reply docs/source/developer/tests.rst:156]: Sounds good!

**Comment by marcoceppi** [comment]
Thanks for the review. I'd like to rebuild this document and make it more concise. I tried to work around the placebo documentation and existing material but looking back over it again I think it would work better to flow something like:

- Writing a functional test
  * Discuss pytest-terraform, placebo record / replay mode with iterative examples
  * Outline how to go from `replay=False` & record to replay
  * test fixture
- Running functional tests
  * Document C7N_FUNCTIONAL variable, highlight how to target just functional (terraform marked) tests
- Migrating existing functional tests
  * Highlight TestClass conversion, other caveats

Thoughts on that as an outline?

File: docs/source/developer/tests.rst [hunk 7]
```diff
@@ -39,67 +39,143 @@ If you need access to Windows you can download a
...
+
+Creating Cloud Resources with Terraform
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Note: These instructions are only for recording data in AWS. For Azure, GCP, or
-other cloud providers, see the corresponding documentation for information on how
-to record data there.
+If a test requires pre-existing cloud resources in order to operate, `pytest-terraform` is the preferred method for creating those resources.
+Pytest Terraform uses `Terraform <https://terraform.io>`_ to repeatably & reliably stand up cloud resources.
+Make sure you have installed terraform and the ``terraform`` command is available in you terminals path.
```
Comment (howbazaar) [inline docs/source/developer/tests.rst:60]: terminal's

File: docs/source/developer/tests.rst [hunk 8]
```diff
@@ -39,67 +39,143 @@ If you need access to Windows you can download a
...
+        test.assertEqual(len(resources), 1)
 
 Now run this test using the following command to generate the placebo data:
 
   .. code-block:: shell
 
-    $ nosetests -s -v tests/path/to/test.py
+    $ pytest tests/path/to/test.py -s -v
 
 This may take a little while as the test is contacting AWS.
```
Comment (howbazaar) [inline docs/source/developer/tests.rst:177]: Earlier you replace 'AWS' with 'the cloud', is it worth doing that here too?

**Review by howbazaar — APPROVED** [review]
On the whole I think it reads nicely and has good examples.

Two minor nits, but otherwise good.

File: docs/source/developer/tests.rst [hunk 9]
```diff
@@ -39,67 +39,143 @@ If you need access to Windows you can download a
...
-~~~~~~~~~~~~~~~~
 
-The ``functional`` decorator marks tests that don't require any pre-existing
-AWS context, and can therefore be run cleanly against live AWS.
+Writing Tests for Cloud Controlled Resources
+--------------------------------------------
 
-To run only the tests decorated by ``functional``:
+Cloud Custodian makes use of flight recording to allow for both functional and unit testing.
+Several libraries are used to perform flight recording and replay.
```
Comment (kapilt) [inline docs/source/developer/tests.rst:47]: would switch this line to 

Each of the provider's uses a separate technique that integrates with the provider sdk to handle flight recording of custodian's api calls. For setting up test scenarios and infrastructure provisioning we use pytest-terraform.

please omit the call out on placebo, its aws specific and by that logic we would have the other provider sdk specific libraries called out. its more important to identify the pattern and our abstraction (record_flight/replay_flight) wrt to custodian api calls, and pytest-terraform for infra setup.

File: docs/source/developer/tests.rst (hunk 7 above)
Comment (kapilt) [reply docs/source/developer/tests.rst:60]: s/terminal/shell PATH

**Review by kapilt — COMMENTED** [review]
some minors, else lgtm, thanks

File: docs/source/developer/tests.rst (hunk 8 above)
Comment (kapilt) [reply docs/source/developer/tests.rst:193]: the example is pretty specific to AWS in this context so the explicit name seems fine.

**Review by kapilt — APPROVED** [review]
thanks, lgtm

**Comment by marcoceppi** [comment]
Thanks for the tweaks, I've sync'd the branch with upstream

**Review by kapilt — APPROVED** [review]
lgtm, thanks

**Pull request merged**